### PR TITLE
Add /invite for locked teams. Fixes #647

### LINKED ISSUE #647: Make a invite command for locked teams.
> That sounds like we'd need a feature to invite people into locked teams, like that no unwanted 'randoms' could join it  - Henningstone

- Something like /invite \<name>
- Detect the team where you are.

### PR DESCRIPTION
**Pull request by Learath2**
This adds /invite as a simple bitmask.

### CODE REVIEW

File: src/game/server/ddracechat.cpp [hunk 1]
```diff
@@ -770,9 +770,115 @@ void CGameContext::ConLockTeam(IConsole::IResult *pResult, void *pUserData)
...
 }
 
+void CGameContext::ConInviteTeam(IConsole::IResult *pResult, void *pUserData)
+{
+	CGameContext *pSelf = (CGameContext *)pUserData;
+	CGameControllerDDRace *pController = (CGameControllerDDRace *)pSelf->m_pController;
+
+	const char *pName = pResult->GetString(0);
+
+
```
Comment (heinrich5991) [inline src/game/server/ddracechat.cpp]: Whiiiiiiitespace.

File: src/game/server/ddracechat.cpp [hunk 2]
```diff
@@ -770,9 +770,115 @@ void CGameContext::ConLockTeam(IConsole::IResult *pResult, void *pUserData)
...
+					return;
+				}
+
+				pController->m_Teams.SetClientInvited(Team, Target, false);
+
+				char aBuf[512];
+				str_format(aBuf, sizeof aBuf, "Your invite to team %d is no longer valid.", Team);
+				pSelf->SendChatTarget(Target, aBuf);
+
+				pSelf->Console()->Print(IConsole::OUTPUT_LEVEL_STANDARD, "invite", "Player has been notified");
```
Comment (heinrich5991) [inline src/game/server/ddracechat.cpp]: Double indent.

File: src/game/server/ddracechat.h [hunk 3]
```diff
@@ -31,6 +31,8 @@ CHAT_COMMAND("rank", "?r[playername]", CFGFLAG_CHAT|CFGFLAG_SERVER, ConRank, thi
 CHAT_COMMAND("rules", "", CFGFLAG_CHAT|CFGFLAG_SERVER, ConRules, this, "Shows the server rules")
 CHAT_COMMAND("team", "?i[id]", CFGFLAG_CHAT|CFGFLAG_SERVER, ConJoinTeam, this, "Lets you join team i (shows your team if left blank)")
 CHAT_COMMAND("lock", "?i['0'|'1']", CFGFLAG_CHAT|CFGFLAG_SERVER, ConLockTeam, this, "Lock team so no-one else can join it")
+CHAT_COMMAND("invite", "s[player name]", CFGFLAG_CHAT|CFGFLAG_SERVER, ConInviteTeam, this, "Invite a person to a locked team")
+CHAT_COMMAND("uninvite", "?s[player name]", CFGFLAG_CHAT|CFGFLAG_SERVER, ConUnInviteTeam, this, "Revoke a persons invitation to a locked team")
```
Comment (heinrich5991) [inline src/game/server/ddracechat.h]: `r` would fit better than `s`, allows whitespace in the name.

File: src/game/server/teams.cpp [hunk 4]
```diff
@@ -646,7 +647,30 @@ void CGameTeams::OnCharacterDeath(int ClientID, int Weapon)
...
 		m_TeamLocked[Team] = Lock;
+	}
+}
+
+void CGameTeams::ResetInvited(int Team)
+{
+	m_Invited[Team] = 0;
+	for (int i = 0; i < MAX_CLIENTS; i++)
+		if(m_Core.Team(i) == Team && GameServer()->m_apPlayers[i])
+			m_Invited[Team] |= 1LL << i;
```
Comment (heinrich5991) [inline src/game/server/teams.cpp]: Why is (if I understand this correctly) everyone from the team automatically invited? What purpose does that code have?

File: src/game/server/teams.h [hunk 5]
```diff
@@ -12,6 +12,7 @@ class CGameTeams
 	bool m_TeeFinished[MAX_CLIENTS];
 	bool m_TeamLocked[MAX_CLIENTS];
 	bool m_IsSaving[MAX_CLIENTS];
+	int64_t m_Invited[MAX_CLIENTS];
```
Comment (heinrich5991) [inline src/game/server/teams.h]: Should probably be `uint64` (without `_t`), we have our own types IIRC.

File: src/game/server/teams.cpp [hunk 6]
```diff
@@ -646,7 +647,30 @@ void CGameTeams::OnCharacterDeath(int ClientID, int Weapon)
...
+}
+
+void CGameTeams::SetClientInvited(int Team, int ClientID, bool Invited)
+{
+	if(Team > TEAM_FLOCK && Team < TEAM_SUPER)
+	{
+		if(Invited)
+			m_Invited[Team] |= 1LL << ClientID;
```
Comment (heinrich5991) [inline src/game/server/teams.cpp]: Here and everywhere with that constant: Shifting into the sign bit is undefined, do we maybe want `1ULL`? Don't know the exact prefixes.

File: src/game/server/teams.h (hunk 5 above)
Comment (Learath2) [inline src/game/server/teams.h]: I've never seen them and a quick grep reveals nothing.

File: src/game/server/teams.cpp (hunk 4 above)
Comment (Learath2) [inline src/game/server/teams.cpp]: I didn't want people getting stuck outside of their teams. I guess could be removed.

File: src/game/server/teams.h (hunk 5 above)
Comment (Henningstone) [inline src/game/server/teams.h]: it's just `int64`, which is a typedef for `long long`

**Comment by heinrich5991** [comment]
Can someone check whether this compiles on Windows? If so, this can be merged.

**Comment by Learath2** [comment]
@Redix checked it.

**Comment by heinrich5991** [comment]
@Redix says this compiles on Windows. Thanks for the pull request, @Learath2.

**Pull request merged**